Add distributed debug mode func to python

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52803 Log nccl debug level in ProcessGroupNCCL
* **#52481 Add distributed debug mode func to python**
used in loss computation
* #52385 [DDP] Enhance warning for find_unused_params
* #52384 [DDP] unittest for when params arent used in backward pass

Adds an API `get_debug_mode` that can be used by distributed package and users to retrieve debug mode. Currently no functionality changes, but wanted to get the bare bones function out and add relevant debug mode logging in follow up diffs.

Detailed docs will be added once functionality is enabled.

Differential Revision: [D26508972](https://our.internmc.facebook.com/intern/diff/D26508972/)

rohan-varma committed:
Add distributed debug mode func to python

Adds an API `get_debug_mode` that can be used by distributed package and users to retrieve debug mode. Currently no functionality changes, but wanted to get the bare bones function out and add relevant debug mode logging in follow up diffs.

Differential Revision: [D26508972](https://our.internmc.facebook.com/intern/diff/D26508972/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/__init__.py:
```diff
@@ -18,6 +29,22 @@ def is_available():
 if is_available() and not torch._C._c10d_init():
     raise RuntimeError("Failed to initialize torch.distributed")
 
+def get_debug_mode():
```
Does this method have to be public? Can we keep it private for now and also make the enum `DistributedDebugMode` private? I'm assuming these are only for internal use and users would only be setting `TORCH_DISTRIBUTED_DEBUG`?

zhaojuanmao replied:
I thought we want OSS users to use the debug mode features as well? it will be helpful for OSS users to debug issues

rohan-varma replied:
I think the reasoning is that PyTorch user themselves won't have a need to call this function or use the enum added here, it will only be used in pytorch code, so we should _ it.

pritamdamania87 commented on torch/distributed/__init__.py:
```diff
@@ -18,6 +29,22 @@ def is_available():
+    Reads the environment variable ``TORCH_DISTRIBUTED_DEBUG`` and returns a
+    ``DistributedDebugMode`` corresponding to the appropriate debug level. The
+    env var ``TORCH_DISTRIBUTED_DEBUG`` must be set to one of "OFF", "INFO", or
+    "DETAIL". Default setting is INFO.
```
Shouldn't default be OFF?

rohan-varma replied:
We can probably do off for now while this is under development, although I think the proposal was to go with INFO by default and verify that it doesn't have a significant perf hit. DETAIL would have things such as the ProcessGroupWrapper that can affect performance. But during development we can stick with OFF, and measure the perf impact of DETAIL when switching over

pritamdamania87 commented on test/distributed/test_c10d.py:
```diff
@@ -4649,6 +4649,29 @@ def test_gloo_barrier_device_ids(self):
+            "INFO": dist.DistributedDebugMode.INFO,
+            "DETAIL": dist.DistributedDebugMode.DETAIL,
+        }
+        invalid_debug_modes = [None, "foo", -1]
+        for invalid_mode in invalid_debug_modes:
+            mapping[invalid_mode] = dist.DistributedDebugMode.INFO
```
Seems like these should raise an error instead of defaulting to INFO. I guess only `None` should default to INFO, but rest (foo, -1) should raise an error.

rohan-varma replied:
This sounds good, although I guess I thought a benefit of the current approach is we'd default to something sensible in these cases, though it might be better to be explicit and error out

rohan-varma replied:
Actually, the `None` doesn't really make sense here since we convert it to string, and "None" should also be an invalid value.

rohan-varma committed:
Update on "Add distributed debug mode func to python"

used in loss computation
* #52385 [DDP] Enhance warning for find_unused_params
* #52384 [DDP] unittest for when params arent used in backward pass

Adds an API `get_debug_mode` that can be used by distributed package and users to retrieve debug mode. Currently no functionality changes, but wanted to get the bare bones function out and add relevant debug mode logging in follow up diffs.

Detailed docs will be added once functionality is enabled.

Differential Revision: [D26508972](https://our.internmc.facebook.com/intern/diff/D26508972/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add distributed debug mode func to python"

used in loss computation
* #52385 [DDP] Enhance warning for find_unused_params
* #52384 [DDP] unittest for when params arent used in backward pass

Adds an API `get_debug_mode` that can be used by distributed package and users to retrieve debug mode. Currently no functionality changes, but wanted to get the bare bones function out and add relevant debug mode logging in follow up diffs.

Detailed docs will be added once functionality is enabled.

Differential Revision: [D26508972](https://our.internmc.facebook.com/intern/diff/D26508972/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add distributed debug mode func to python"

used in loss computation
* #52385 [DDP] Enhance warning for find_unused_params
* #52384 [DDP] unittest for when params arent used in backward pass

Adds an API `get_debug_mode` that can be used by distributed package and users to retrieve debug mode. Currently no functionality changes, but wanted to get the bare bones function out and add relevant debug mode logging in follow up diffs.

Detailed docs will be added once functionality is enabled.

Differential Revision: [D26508972](https://our.internmc.facebook.com/intern/diff/D26508972/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add distributed debug mode func to python"

used in loss computation
* #52385 [DDP] Enhance warning for find_unused_params
* #52384 [DDP] unittest for when params arent used in backward pass

Adds an API `get_debug_mode` that can be used by distributed package and users to retrieve debug mode. Currently no functionality changes, but wanted to get the bare bones function out and add relevant debug mode logging in follow up diffs.

Detailed docs will be added once functionality is enabled.

Differential Revision: [D26508972](https://our.internmc.facebook.com/intern/diff/D26508972/)

[ghstack-poisoned]
